File: packages_py/runtime_template_resolver/src/runtime_template_resolver/resolver.py
from typing import Any
from .path_parser import parse_path

class SecurityError(Exception):
    pass

def resolve_path(obj: Any, path: str) -> Any:
    """
    Safely resolve a value from a nested object using a path string.
    Does not use eval(). Prevents access to private attributes (starting with _).
    """
    segments = parse_path(path)
    current = obj

    for segment in segments:
        if current is None:
            return None

        # Prevent attribute injection/private access
        if segment.startswith('_'):
             raise SecurityError(f"Unsafe path segment: {segment}")

        if isinstance(current, dict):
            current = current.get(segment)
        elif isinstance(current, (list, tuple)):
            if segment.isdigit():
                idx = int(segment)
                current = current[idx] if idx < len(current) else None
            else:
                return None
        elif hasattr(current, segment):
            current = getattr(current, segment)
        else:
            return None

    return current

File: packages_mjs/runtime-template-resolver/src/resolver.ts
import { parsePath } from './path-parser.js';

export class SecurityError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'SecurityError';
    }
}

export function resolvePath(obj: unknown, path: string): unknown {
    const segments = parsePath(path);
    let current: any = obj;

    for (const segment of segments) {
        if (current === null || current === undefined) {
            return undefined;
        }

        // Prevent prototype pollution / unsafe access
        if (segment === '__proto__' || segment === 'constructor' || segment === 'prototype') {
            throw new SecurityError(`Unsafe path segment: ${segment}`);
        }

        if (typeof current !== 'object') {
            return undefined;
        }

        current = current[segment];
    }

    return current;
}
